services/retention/service.go

```diff
@@ -80,83 +108,135 @@ func (s *Service) run() {
+					deletedShardIDs[sh.ID] = newDeletionInfo(d.Name, r.Name, sh)
+				}
 			}
-			deletedShardIDs := make(map[uint64]deletionInfo)
-
-			// Mark down if an error occurred during this function so we can inform the
-			// user that we will try again on the next interval.
-			// Without the message, they may see the error message and assume they
-			// have to do it manually.
-			var retryNeeded bool
-			dbs := s.MetaClient.Databases()
-			for _, d := range dbs {
-				for _, r := range d.RetentionPolicies {
-					// Build list of already deleted shards.
-					for _, g := range r.DeletedShardGroups() {
-						for _, sh := range g.Shards {
-							deletedShardIDs[sh.ID] = deletionInfo{db: d.Name, rp: r.Name}
-						}
-					}
-
-					// Determine all shards that have expired and need to be deleted.
-					for _, g := range r.ExpiredShardGroups(time.Now().UTC()) {
-						if err := s.MetaClient.DeleteShardGroup(d.Name, r.Name, g.ID); err != nil {
-							log.Info("Failed to delete shard group",
-								logger.Database(d.Name),
-								logger.ShardGroup(g.ID),
-								logger.RetentionPolicy(r.Name),
-								zap.Error(err))
-							retryNeeded = true
-							continue
-						}
-
-						log.Info("Deleted shard group",
-							logger.Database(d.Name),
-							logger.ShardGroup(g.ID),
-							logger.RetentionPolicy(r.Name))
-
-						// Store all the shard IDs that may possibly need to be removed locally.
-						for _, sh := range g.Shards {
-							deletedShardIDs[sh.ID] = deletionInfo{db: d.Name, rp: r.Name}
-						}
-					}
+
+			// Determine all shards that have expired and need to be deleted.
+			for _, g := range r.ExpiredShardGroups(time.Now().UTC()) {
+				if err := s.MetaClient.DeleteShardGroup(d.Name, r.Name, g.ID); err != nil {
+					log.Info("Failed to delete shard group",
+						logger.Database(d.Name),
+						logger.ShardGroup(g.ID),
+						logger.RetentionPolicy(r.Name),
+						zap.Error(err))
+					retryNeeded = true
+					continue
+				}
+
+				log.Info("Deleted shard group",
+					logger.Database(d.Name),
+					logger.ShardGroup(g.ID),
+					logger.RetentionPolicy(r.Name))
+
+				// Store all the shard IDs that may possibly need to be removed locally.
+				for _, sh := range g.Shards {
+					deletedShardIDs[sh.ID] = newDeletionInfo(d.Name, r.Name, sh)
 				}
 			}
+		}
+	}
 
-			// Remove shards if we store them locally
-			for _, id := range s.TSDBStore.ShardIDs() {
-				if info, ok := deletedShardIDs[id]; ok {
-					if err := s.TSDBStore.DeleteShard(id); err != nil {
-						log.Info("Failed to delete shard",
-							logger.Database(info.db),
-							logger.Shard(id),
-							logger.RetentionPolicy(info.rp),
-							zap.Error(err))
-						retryNeeded = true
-						continue
-					}
-					log.Info("Deleted shard",
+	// Remove shards if we store them locally
+	for _, id := range s.TSDBStore.ShardIDs() {
+		if info, ok := deletedShardIDs[id]; ok {
+			delete(deletedShardIDs, id)
+			if err := s.TSDBStore.DeleteShard(id); err != nil {
```

Comment:
One more change - what do you think about printing one more message before calling `DeleteShard`: 
- `attempting deletion of shard X, database Y, retention policy Z`

We have no visibility into why the retention service hangs for so long, and this might give us more clues; we can look at the size and cardinality of a shard that prints the message and then never prints a success or failure message.